src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -66,20 +66,64 @@ private void BeforeExecuteCommand(TypeCharCommandArgs args, CommandExecutionCont
             }
 
             var caret = caretOpt.Value;
-
             var document = caret.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
             if (document == null)
             {
                 return;
             }
 
-            // on the UI thread
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
             var root = document.GetSyntaxRootSynchronously(executionContext.OperationContext.UserCancellationToken);
+
+            var currentNode = GetStartingNode(document, root, caret);
+            MoveCaretToSemicolonPosition(args, document, root, caret, syntaxFacts, currentNode, isInsideDelimiters: false);
+        }
+
+        private void MoveCaretToSemicolonPosition(TypeCharCommandArgs args, Document document, SyntaxNode root, SnapshotPoint caret, ISyntaxFactsService syntaxFacts, SyntaxNode currentNode, bool isInsideDelimiters)
```

Comment:
📝 File a follow-up issue to convert this method from recursive to iterative. We don't want to trigger a stack overflow when semicolon is typed within very very very very deeply nested parentheses.

Reply:
This is filed, #35156